src/test/spec/unified_runner/operation.rs

```diff
@@ -0,0 +1,801 @@
+use std::{collections::HashMap, fmt::Debug, ops::Deref};
+
+use async_trait::async_trait;
+use futures::stream::TryStreamExt;
+use serde::Deserialize;
+
+use super::{Entity, ExpectError, Operation};
+
+use crate::{
+    bson::{doc, Bson, Deserializer as BsonDeserializer, Document},
+    error::Result,
+    options::{
+        AggregateOptions,
+        CountOptions,
+        DeleteOptions,
+        DistinctOptions,
+        EstimatedDocumentCountOptions,
+        FindOneAndDeleteOptions,
+        FindOneAndReplaceOptions,
+        FindOneAndUpdateOptions,
+        FindOneOptions,
+        FindOptions,
+        InsertManyOptions,
+        InsertOneOptions,
+        ListCollectionsOptions,
+        ListDatabasesOptions,
+        ReplaceOptions,
+        UpdateModifications,
+        UpdateOptions,
+    },
+    test::util::EventClient,
+    Collection,
+    Database,
+};
+
+#[async_trait]
+pub trait TestOperation: Debug {
+    /// The command names to monitor as part of this test.
+    fn command_names(&self) -> &[&str];
+
+    // The linked issue causes a warning that cannot be suppressed when providing a default
+    // implementation for these functions.
+    // <https://github.com/rust-lang/rust/issues/51443>
+    async fn execute_on_collection(&self, collection: &Collection) -> Result<Option<Bson>>;
+
+    async fn execute_on_client(&self, client: &EventClient) -> Result<Option<Bson>>;
+
+    async fn execute_on_database(&self, database: &Database) -> Result<Option<Bson>>;
+}
+
+#[derive(Debug)]
+pub struct EntityOperation {
+    operation: Box<dyn TestOperation>,
+    pub name: String,
+    pub object: String,
```

Comment:
That won't quite work since the `object` field is the name of a specific entity in the entity map to call the operation on (e.g. something like "client1" or "database5").

Reply:
What about the following:
```rust
enum OperationObject {
    Entity { id: String },
    TestRunner
}
```
This could ensure we don't accidentally use the "testRunner" string as an entity id.